### PR TITLE
Merge/2.12.x to 2.13.x 2017020 [ci: last-only]

### PR DESCRIPTION
**Pull request by szeiger**
Needs to be reviewed by someone who knows the sbt / REPL interaction. Everything is different in 2.13.x because of the REPL refactoring. @dwijnand? @som-snytt?

```
$ git checkout -b merge/2.12.x-to-2.13.x-20170720 scala/2.13.x

$ export MB=$(git merge-base HEAD scala/2.12.x)

$ git log --graph --oneline --decorate $MB..scala/2.12.x
*   de9c93a55d (scala/2.12.x) Merge pull request #6000 from retronym/topic/minimal-scalacheck-fix
|\
| * 18721ead44 [nomerge] Cleanup threads created by parallel collections tests
| * 71f9b43c9e Failing scalacheck tests fail build again
|/
*   ba9dacc206 Merge pull request #5998 from retronym/review/5918
|\
| * d64cf9ef23 PipedSource should join Source thread to wait the exit value.
*   0dc2c3dc60 Merge pull request #5655 from som-snytt/issue/10102
|\
| * 4347dd9a4e SI-10102 REPL ignores sbt init errors
* 085be56899 Merge pull request #5977 from lrytz/t10334-bridge
* f1b610c46e Add a bridge to LMF-generated lambdas with the instantiated signature

$ git merge 71f9b43c9e
CONFLICT (modify/delete): src/repl/scala/tools/nsc/interpreter/ReplReporter.scala deleted in HEAD and modified in 71f9b43c9e. Version 71f9b43c9e of src/repl/scala/tools/nsc/interpreter/ReplReporter.scala left in tree.
Auto-merging src/repl/scala/tools/nsc/interpreter/IMain.scala
CONFLICT (content): Merge conflict in src/repl/scala/tools/nsc/interpreter/IMain.scala
CONFLICT (modify/delete): src/repl/scala/tools/nsc/InterpreterLoop.scala deleted in HEAD and modified in 71f9b43c9e. Version 71f9b43c9e of src/repl/scala/tools/nsc/InterpreterLoop.scala left in tree.
Auto-merging src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala
CONFLICT (content): Merge conflict in src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala
Auto-merging src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala
Auto-merging build.sbt
Automatic merge failed; fix conflicts and then commit the result.

# conflicts from https://github.com/scala/scala/pull/5655 due to REPL refactoring

$ git merge 18721ead44 -s ours
Merge made by the 'ours' strategy.

$ git merge scala/2.12.x
Already up-to-date!
Merge made by the 'recursive' strategy.
```

### CODE REVIEW

**Comment by lrytz** [comment]
From now on we should only do backports for repl changes.

**Comment by lrytz** [comment]
If it makes it easier, you could also merge the repl PR with `-s ours` and do the equivalent changes in 2.13.x by hand, but now you're already done anyway, so it's just about reviewing.

**Comment by som-snytt** [comment]
REPL tweaks on 2.12 should all be `[nomerge]`?  These won't be backports because 2.13 will not need the tweaks, because it will be refactored to a high gloss finish.

There's still

https://github.com/scala/scala/pull/5993
https://github.com/scala/scala/pull/5657

File: src/repl-frontend/scala/tools/nsc/Interpreter.scala [hunk 1]
```diff
@@ -32,7 +32,7 @@ class InterpreterLoop {
   private var compilerSettings: Settings = _
   private var parentClassLoader: Option[ClassLoader] = _
 
-  @deprecated("Mainly used for passing in settings.", "2.13.0-M2")
+  // @deprecated("Mainly used for passing in settings.", "2.13.0-M2")
```
Comment (lrytz) [inline src/repl-frontend/scala/tools/nsc/Interpreter.scala]: why did you un-deprecate this one?

File: src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala [hunk 2]
```diff
@@ -103,6 +103,9 @@ class ILoop(config: ShellConfig, inOverride: BufferedReader = null,
 
   def welcome = enversion(welcomeString)
 
+  // ignore silent sbt errors on init
+  protected def isSbt: Boolean = false
```
Comment (lrytz) [inline src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala]: `isSbt` is never `true`, so this doesn't seem to do anything

**Review by lrytz — CHANGES_REQUESTED** [review]
@szeiger I think it's better to merge 0dc2c3dc60 with `-s ours`. I'd say it's fine to delay the fix for scala/bug#10102 to 2.13.0-M3.

It's really hard to merge any non-trivial repl change forward, so IMO all repl changes that go to 2.12.x should be `[nomerge]` and come with an equivalent PR to 2.13.x (if applicable).

File: src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala (hunk 2 above)
Comment (szeiger) [reply src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala]: Indeed, in 2.12.x it was overridden in InterpreterLoop but InterpreterLoop no longer extends ILoop.

**Comment by szeiger** [comment]
> If it makes it easier, you could also merge the repl PR with -s ours and do the equivalent changes in 2.13.x by hand

That's pretty much what I did during the merge. The branches are so different that I had to recreate most changes from #5655.  I'll redo the merge without them. It doesn't seem to be much sense like this in 2.13.x.

**Comment by szeiger** [comment]
New version without #5655, no conflicts:

```
$ git checkout -b merge/2.12.x-to-2.13.x-20170720 scala/2.13.x

$ export MB=$(git merge-base HEAD scala/2.12.x)

$ git log --graph --oneline --decorate $MB..scala/2.12.x
*   de9c93a55d (scala/2.12.x) Merge pull request #6000 from retronym/topic/minimal-scalacheck-fix
|\
| * 18721ead44 [nomerge] Cleanup threads created by parallel collections tests
| * 71f9b43c9e Failing scalacheck tests fail build again
|/
*   ba9dacc206 Merge pull request #5998 from retronym/review/5918
|\
| * d64cf9ef23 PipedSource should join Source thread to wait the exit value.
*   0dc2c3dc60 Merge pull request #5655 from som-snytt/issue/10102
|\
| * 4347dd9a4e SI-10102 REPL ignores sbt init errors
* 085be56899 Merge pull request #5977 from lrytz/t10334-bridge
* f1b610c46e Add a bridge to LMF-generated lambdas with the instantiated signature

$ git merge 085be56899
Auto-merging src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala
Merge made by the 'recursive' strategy.
 src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala | 10 +++++++---
 test/files/run/t10334.scala                                     | 70 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 test/files/run/t10334b/A.java                                   |  4 ++++
 test/files/run/t10334b/Test.scala                               | 21 +++++++++++++++++++++
 4 files changed, 102 insertions(+), 3 deletions(-)
 create mode 100644 test/files/run/t10334.scala
 create mode 100644 test/files/run/t10334b/A.java
 create mode 100644 test/files/run/t10334b/Test.scala

$ git merge 0dc2c3dc60 -s ours
Merge made by the 'ours' strategy.

$ git merge 71f9b43c9e
Auto-merging build.sbt
Merge made by the 'recursive' strategy.
 build.sbt                                                      |  2 +-
 src/library/scala/sys/process/ProcessImpl.scala                |  1 +
 src/reflect/scala/reflect/internal/Symbols.scala               | 18 ++++++++----------
 test/junit/scala/sys/process/PipedProcessTest.scala            | 19 +++++++++++++++++++
 test/scalacheck/scala/tools/nsc/scaladoc/HtmlFactoryTest.scala | 59 +++++++++++++++++++++++++----------------------------------
 test/scalacheck/scala/tools/nsc/scaladoc/IndexScriptTest.scala | 21 ++++++---------------
 test/scalacheck/scala/tools/nsc/scaladoc/SettingsUtil.scala    | 30 ++++++++++++++++++++++++++++++
 test/scaladoc/resources/t4421.scala                            |  2 +-
 test/scaladoc/resources/t4507.scala                            |  2 +-
 test/scaladoc/resources/t4589.scala                            |  2 +-
 test/scaladoc/resources/t4715.scala                            |  2 +-
 test/scaladoc/resources/t4898.scala                            |  2 +-
 test/scaladoc/resources/t5054_q1.scala                         |  2 +-
 test/scaladoc/resources/t5054_q2.scala                         |  2 +-
 test/scaladoc/resources/t5054_q3.scala                         |  2 +-
 test/scaladoc/resources/t5054_q4.scala                         |  2 +-
 test/scaladoc/resources/t5054_q5.scala                         |  2 +-
 test/scaladoc/resources/t5054_q6.scala                         |  2 +-
 test/scaladoc/resources/t5054_q7.scala                         |  2 +-
 test/scaladoc/resources/t5287.scala                            |  6 +++---
 20 files changed, 105 insertions(+), 75 deletions(-)
 create mode 100644 test/scalacheck/scala/tools/nsc/scaladoc/SettingsUtil.scala

 $ git merge 18721ead44 -s ours
Merge made by the 'ours' strategy.

$ git merge scala/2.12.x
Already up-to-date!
Merge made by the 'recursive' strategy.
```

**Review by lrytz — APPROVED** [review]

**Pull request merged**